ci/remote-watch.py

```diff
@@ -0,0 +1,247 @@
+        nonsignals.append(signal.CTRL_C_EVENT)
+        nonsignals.append(signal.CTRL_BREAK_EVENT)
+    sig = None
+    try:
+        attempts = [
+            signal.SIGINT if is_posix else signal.CTRL_C_EVENT,
+            signal.SIGTERM if is_posix else signal.CTRL_BREAK_EVENT,
+            signal.SIGKILL if is_posix else signal.SIGTERM,
+        ]
+        for i, sig in enumerate(attempts):
+            tsleep = sleeps[i] if i < len(sleeps) else None
+            if pid == 0 or sig not in nonsignals:
+                logger.info("Attempting os.kill(%s, %s)...", pid, sig)
+                old_handler = None
+                if pid == 0:  # Ignore our own signal
+                    try:
+                        old_handler = signal.signal(sig, signal.SIG_IGN)
+                    except (OSError, RuntimeError, ValueError):
+                        pass  # Some signals can't be ignored; just continue
+                try:
+                    os.kill(pid, sig)
+                except OSError as ex:
+                    if sys.platform == "win32" and ex.winerror == 87:
+                        pass  # This can sometimes happen; just ignore it
+                    else:
+                        raise
+                finally:
+                    if pid == 0 and old_handler is not None:  # Restore handler
+                        signal.signal(sig, old_handler)
+                time.sleep(tsleep)
+    except OSError as ex:
+        if ex.errno not in (errno.EBADF, errno.ESRCH):
+            raise
+        logger.error("os.kill(%r, %r) error %s: %s", pid, sig, ex.errno,
+                     ex.strerror)
+        result = ex.errno
+    return result
+
+
+def monitor(server_poll_interval=None, sleeps=None):
+    if sleeps is None:
+        sleeps = [5, 15]
+    if not server_poll_interval:
+        server_poll_interval = 30
+    ci = get_current_ci() or ""
+    (ref, remote, expected_sha) = git_branch_info()
+    expected_line = "{}\t{}".format(expected_sha, ref)
+    commit_msg = git("show", "-s", "--format=%B", "HEAD^-")
+    keep_alive = False
+    for match in re.findall("(?m)^([# ]*CI_KEEP_ALIVE(:(.*))?)$", commit_msg):
+        if not match[2].strip():
+            keep_alive = True
+        else:
+            for ci_name in match[2].split(","):
+                if ci_name.strip().lower() == ci.lower():
+                    keep_alive = True
+    logger.info("Monitoring %s (%s) for changes in %s every %s seconds: %s",
+                remote, git("ls-remote", "--get-url", remote),
+                ref, server_poll_interval, expected_line)
+    prev_ignored_line = None
+    tstart = default_timer()
+    tprev = None
+    terminate = False
+    while True:
+        to_wait = 0
+        if tprev is not None:
+            to_wait = max(tprev + server_poll_interval - default_timer(), 0)
+
+        # Try to increase waiting time gradually, for better responsiveness
+        to_wait = min(to_wait, max(default_timer() - tstart, 5.0))
```

Comment:
why? just checking every 5min should suffice.

Reply:
5 minutes might not seem like much for the core code, but it's way too painfully slow when you're iterating on the CI setup code, which is in the beginning. So I'd really rather keep this unless we see it causes a problem -- I've used this code before and it's worked really well for me.

Reply:
2 minutes then. The goal here is to make is readable for developers with minimal context. For we can also just use a static schedule:
[0.5, 0.5, 1, 1, 2, 5, 5, ....]

Reply:
But imagine we have like 16 jobs polling every 2 minutes each... if each build is 1 hour that's going to hit GitHub 480 times during that timespan. I can switch to a static schedule if you find that easier to understand though.

Reply:
I just realized I didn't even think about this fact when I initially wrote this. I'll need to add that logic.